### PR TITLE
Add Transition logic for WaitingForPushDeviceDetails State

### PR DESCRIPTION
**Pull request by marto83**

### CODE REVIEW

File: src/IO.Ably.Shared/Push/PushAdmin.cs [hunk 1]
```diff
@@ -30,6 +30,27 @@ internal PushAdmin(AblyRest restClient, ILogger logger)
...
+        {
+            // TODO: Add validation.
+            // TODO: Add fullwait parameter.
+            var request = _restClient.CreateRequest("/push/deviceRegistrations/", HttpMethod.Post);
+            request.PostData = details;
+            var result = await _restClient.ExecuteRequest<LocalDevice>(request);
+
+            if (result == null)
+            {
+                throw new AblyException("Failed to register device", 40000);
```
Comment (tomkirbygreen) [inline src/IO.Ably.Shared/Push/PushAdmin.cs:48]: Magic number 40,000. Can we have a symbolic?

File: src/IO.Ably.Shared/Push/ActivationStateMachine.States.cs [hunk 2]
```diff
@@ -80,6 +81,76 @@ public WaitingForPushDeviceDetails(ActivationStateMachine machine)
...
+                        var ably = Machine._restClient; // TODO: Check if there is an instance when Ably is not set. In which case Java set queues GettingDeviceRegistrationFailed
+
+                        try
+                        {
+                            var registeredDevice = await ably.Push.Admin.RegisterDevice(device);
+                            var deviceIdentityToken = registeredDevice.DeviceIdentityToken;
+                            if (deviceIdentityToken.IsEmpty())
+                            {
+                                // TODO: Log
+                                _ = Machine.HandleEvent(new GettingDeviceRegistrationFailed(new ErrorInfo("Invalid deviceIdentityToken in response", 40000, HttpStatusCode.BadRequest)));
```
Comment (tomkirbygreen) [inline src/IO.Ably.Shared/Push/ActivationStateMachine.States.cs:109]: Magic number 40,000. Consider using a symbolic?

**Review by tomkirbygreen — APPROVED** [review]
Just wondering about the `40000` magic number ...

File: src/IO.Ably.Shared/Push/PushAdmin.cs (hunk 1 above)
Comment (paddybyers) [reply src/IO.Ably.Shared/Push/PushAdmin.cs:48]: If the HTTP request failed, then we need this error response to contain information about that error, not just a fixed, non-specific error code

File: src/IO.Ably.Shared/Push/ActivationStateMachine.States.cs (hunk 2 above)
Comment (paddybyers) [reply src/IO.Ably.Shared/Push/ActivationStateMachine.States.cs:109]: So if I'm reading this correctly, we made a successful device registration request, but the response didn't include a `deviceIdentityToken`. I think that's an internal error - ie something that we think would only happen as a result of something being broken. In that case, it's a 5xx error, not 4xx. And I think there should be a specific error code for it.

File: src/IO.Ably.Shared/Push/PushAdmin.cs (hunk 1 above)
Comment (marto83) [reply src/IO.Ably.Shared/Push/PushAdmin.cs:48]: You both are right. I'm only hooking things up to get to a point where I can complete a happy path integration. Thus the numerous TODOs and fixed error codes. I'm going to add a TODO label here as well to make sure it doesn't get dropped.

File: src/IO.Ably.Shared/Push/ActivationStateMachine.States.cs (hunk 2 above)
Comment (marto83) [reply src/IO.Ably.Shared/Push/ActivationStateMachine.States.cs:109]: Good point Paddy. I'm porting the code from the java library. Do you want me to create issues and create PRs there as well?

**Pull request merged**